lib/elastic_apm.rb

```diff
@@ -16,142 +17,237 @@
+    # @param context [Span::Context] Context information about the span
+    # @yield [Span] Optional block encapsulating span
+    # @return [Span] Unless block given
+    def span(name, type = nil, context: nil, include_stacktrace: true, &block)
+      return (block_given? ? yield : nil) unless agent
+
+      if block_given?
+        with_span(
+          name,
+          type,
+          context:
+          context,
+          include_stacktrace: include_stacktrace,
+          &block
+        )
+      else
+        start_span(
+          name,
+          type,
+          context: context,
+          include_stacktrace: include_stacktrace
+        )
+      end
+    end
+    # rubocop:enable Metrics/MethodLength
 
-  ### Context
+    deprecate :span, :with_span
 
-  # Set a _tag_ value for the current transaction
-  #
-  # @param key [String,Symbol] A key
-  # @param value [Object] A value (will be converted to string)
-  # @return [Object] The given value
-  def self.set_tag(key, value)
-    agent && agent.set_tag(key, value)
-  end
+    def start_span(name, type = nil, context: nil, include_stacktrace: true)
+      agent&.start_span(
+        name,
+        type,
+        context: context,
+        backtrace: include_stacktrace ? caller : nil
+      )
+    end
 
-  # Provide further context for the current transaction
-  #
-  # @param custom [Hash] A hash with custom information. Can be nested.
-  # @return [Hash] The current custom context
-  def self.set_custom_context(custom)
-    agent && agent.set_custom_context(custom)
-  end
+    def end_span
+      agent&.end_span
+    end
 
-  # Provide a user to the current transaction
-  #
-  # @param user [Object] An object representing a user
-  # @return [Object] Given user
-  def self.set_user(user)
-    agent && agent.set_user(user)
-  end
+    # rubocop:disable Metrics/MethodLength
+    def with_span(name, type = nil, context: nil, include_stacktrace: true)
+      unless block_given?
+        raise ArgumentError,
+          'expected a block. Do you want `start_span\' instead?'
+      end
+
+      return yield nil unless agent
+
+      begin
+        span =
+          start_span(
+            name, type, context: context, include_stacktrace: include_stacktrace
+          )
+        yield span
+      ensure
+        end_span
+      end
+    end
+    # rubocop:enable Metrics/MethodLength
+
+    # Build a [Context] from a Rack `env`. The context may include information
+    # about the request, response, current user and more
+    #
+    # @param rack_env [Rack::Env] A Rack env
+    # @return [Context] The built context
+    def build_context(rack_env)
+      agent&.build_context(rack_env)
+    end
+
+    ### Errors
+
+    # Report and exception to APM
+    #
+    # @param exception [Exception] The exception
+    # @param handled [Boolean] Whether the exception was rescued
+    # @return [Error] The generated [Error]
+    def report(exception, handled: true)
```

Comment:
Did you think about also aligning this with `span` and `transaction`, e.g. use `error`, and `error_message`?

Reply:
I probably should, yeah. Even though I like my names better 😀